File: src/main/java/edu/cornell/cs/apl/viaduct/security/solver/RightHandTerm.java
package edu.cornell.cs.apl.viaduct.security.solver;

import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;

/** Terms that can appear on the right-hand side of constraints. */
public interface RightHandTerm<A> extends ConstraintTerm<A> {
  /**
   * Return an edge that captures the operation preformed by this term. In the constraint graph,
   * this will become an incoming edge of {@link #getNode()}.
   */
  DataFlowEdge<A> getInEdge();
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/solver/RightHandTerm.kt
package io.github.apl_cornell.viaduct.algebra.solver

import io.github.apl_cornell.viaduct.algebra.HeytingAlgebra
import io.github.apl_cornell.viaduct.util.dataflow.DataFlowEdge

/** Terms that can appear on the right-hand side of constraints. */
interface RightHandTerm<A : HeytingAlgebra<A>> : ConstraintTerm<A> {
    /**
     * Return an edge that captures the operation preformed by this term. In the constraint graph,
     * this will become an outgoing edge of [node].
     */
    val outEdge: DataFlowEdge<A>
}
